crates/rbuilder/src/backtest/fetch/csv.rs

```diff
@@ -0,0 +1,113 @@
+use crate::primitives::Order;
+use crate::{
+    backtest::{
+        fetch::datasource::{BlockRef, DataSource},
+        OrdersWithTimestamp,
+    },
+    primitives::{Bundle, TransactionSignedEcRecoveredWithBlobs},
+};
+use alloy_rlp::Decodable;
+use async_trait::async_trait;
+use csv::Reader;
+use eyre::Context;
+use reth::primitives::TransactionSignedEcRecovered;
+use std::{collections::HashMap, fs::File, path::PathBuf};
+use tracing::trace;
+use uuid::Uuid;
+
+#[derive(Debug, Clone)]
+pub struct CSVDatasource {
+    batches: HashMap<u64, Vec<TransactionSignedEcRecovered>>,
+}
+
+impl CSVDatasource {
+    pub fn new(filename: impl Into<PathBuf>) -> eyre::Result<Self> {
+        let batches = Self::load_transactions_from_csv(filename.into())?;
+        Ok(Self { batches })
+    }
+
+    fn load_transactions_from_csv(
+        filename: PathBuf,
+    ) -> eyre::Result<HashMap<u64, Vec<TransactionSignedEcRecovered>>> {
+        let file = File::open(&filename)
+            .wrap_err_with(|| format!("Failed to open file: {}", filename.display()))?;
+        let mut reader = Reader::from_reader(file);
+        let mut batches: HashMap<u64, Vec<TransactionSignedEcRecovered>> = HashMap::new();
+
+        for result in reader.records() {
+            let record = result?;
+            if record.len() != 2 {
+                return Err(eyre::eyre!("Invalid CSV format"));
+            }
+
+            let batch_number: u64 = record[0].parse()?;
+            let rlp_hex = &record[1];
+            let rlp_bytes = hex::decode(rlp_hex)?;
+            let tx = TransactionSignedEcRecovered::decode(&mut &rlp_bytes[..])?;
+
+            batches.entry(batch_number % 10).or_default().push(tx);
+        }
+
+        Ok(batches)
+    }
+}
+
+#[async_trait]
+impl DataSource for CSVDatasource {
+    async fn get_orders(&self, block: BlockRef) -> eyre::Result<Vec<OrdersWithTimestamp>> {
+        // The csv datasource is one with 10 batches, where batch is a list of transactions
+        // Since we don't have full "real" blocks, we'll just use the block number to determine the batch
+        // Thus the usage of mod 10 is just to determine the batch number that we get transactions from, e.g. block 100 corresponds to 0, 101 to 1, 109 to 9, etc.
+        let batch_number = block.block_number % 10;
```

Comment:
this 10 should be a constant.
Maybe also some function like batch_number_from_u64() to isolate the % 10?